### ISSUE TITLE
java error

### ISSUE THREAD

**Issue by hurleyLi** [frazer-lab/cluster#158]
Hi @tatarsky , 

I had the following error when I run a java script

```
java -jar /frazer01/projects/CARDIPS/pipeline/Hi-C/script/repo/helpers/Juicebox_V1.5_customized.jar pre -f /frazer01/projects/CARDIPS/pipeline/Hi-C/script/repo/helpers/MboI.eachChr.1based.sites /frazer01/projects/CARDIPS/pipeline/Hi-C/merged/CM/CM.merged_nodups.filtered.inter.intra_longRange.txt /frazer01/projects/CARDIPS/pipeline/Hi-C/merged/CM/output /software/bedtools-2.25.0/genomes/human.hg19.genome 


Skipping 
Start preprocess
Writing header
Writing body
Oct 22, 2016 8:50:21 PM java.util.prefs.FileSystemPreferences checkLockFile0ErrorCode
WARNING: Could not lock User prefs.  Unix error code 38.
Oct 22, 2016 8:50:21 PM java.util.prefs.FileSystemPreferences syncWorld
WARNING: Couldn't flush user prefs: java.util.prefs.BackingStoreException: Couldn't get file lock.
```

Weird thing is this error only occurred when I run it on a compute node or submit the job to the cluster, but not when I run it on the flh1. 

I found a relevant thread here
http://stackoverflow.com/questions/23960451/java-system-preferences-under-different-users-in-linux
I tried to applied the solution, but the error still occurred. 
Do you have any idea why this might happen?

Thanks
Hurley

**Comment by billgreenwald** [comment]
Did you put the solution from the stack overflow into your job script or your bash rc?

**Comment by tatarsky** [comment]
I will look at this on Monday but to prevent excessive debug effort on a Sunday the phrase "Couldn't get file lock." may be the clue. 

We do NOT run Lustre with file locking support on the _nodes_ but have enabled it on the head nodes (Lustre is mounted with `localflock`)  to support Mysql. That is a "fake" flock that doesn't really communicate the lock to all the other clients. (We don't care for Mysql storing its tables on Lustre running on one head only)

Locking in clustered filesystems is a longer topic and there a few options that depending on WHY your code is locking we may need more than the "localflock" option and accept any metadata performance penalties for needing it to be consistent to all nodes.

You might also see if you can simply tell your Java code to put whatever it wants to lock not on Lustre. (remember clearly /home = Lustre) 

Aka in /scratch or /tmp or something. Ala @billgreenwald suggestion above perhaps drop them something like `export JAVA_OPTS="-Djava.util.prefs.systemRoot=/scratch/hel070/.java ...`.

This assume of course those prefs  don't need to be shared by all the threads and you'd have to `mkdir` that or something.

Again, I will debug more with you Monday including we can offline a node and remount Lustre with the localflock and/or the more distributed option we do not run for metadata performance reasons.

**Comment by tatarsky** [comment]
I remounted `fl-n-1-1` lustre with the localflock (non-distributed version of locking). Please ssh manually to fl-n-1-1 and verify that changes the behavior of your Java item when convenient. We'll discuss next steps if it does and debug further if it doesn't.

**Comment by hurleyLi** [comment]
Just checked, the error is gone on fl-n-1-1.

**Comment by tatarsky** [comment]
So. Boils down here to then "do you know if your code requires cross all-node consistent locks?"

Aka if you lock a file on say node 1 all the other nodes that if they check the status of that lock can see the file locked.

Or is "local lock" good enough? (means only node 1 processes will see that a file is locked when a lock request is issued on node 1)

The more elaborate locking (all nodes can see it) does have performance penalties as it adds additional metadata server information exchange IIRC. I do not recall the class of performance penalty but its non-zero.

Door number three, can you tell your code "we do not care about locks?"

If you need either of the locking methods (local or global) it requires I umount/remount lustre on the nodes which I can schedule once you mull the above.

**Comment by hurleyLi** [comment]
Solved the problem by specifying the locking files (or whatever it's called) on `/scratch/hel070`. I specified the locking file position to `/home/hel070` before, and that might be why it was not working. Also, it looks like I have to specify the options in the command, rather than doing the `export JAVA_OPTS=`

```
java -Djava.util.prefs.systemRoot=/scratch/hel070/.java -Djava.util.prefs.userRoot=/scratch/hel070
```

The command will actually make a folder `/scratch/hel070/.java/.userPrefs`

Thanks guys, cheers!

Hurley
